src/sentry/services/hybrid_cloud/__init__.py

```diff
@@ -0,0 +1,319 @@
+    _singleton: Dict[SiloMode, ServiceInterface]
+
+    def __init__(self, mapping: Mapping[SiloMode, Type[ServiceInterface]]):
+        self._constructors = mapping
+        self._singleton = {}
+
+    @contextlib.contextmanager
+    def with_replacement(self, service: Optional[ServiceInterface], silo_mode: SiloMode):
+        prev = self._singleton
+        self.close()
+
+        if service:
+            self._singleton[silo_mode] = service
+            yield
+        else:
+            yield
+        self.close()
+        self._singleton = prev
+
+    def __getattr__(self, item: str):
+        cur_mode = SiloMode.get_current_mode()
+        if impl := self._singleton.get(cur_mode, None):
+            return getattr(impl, item)
+        if Con := self._constructors.get(cur_mode, None):
+            self.close()
+            return getattr(self._singleton.setdefault(cur_mode, Con()), item)
+
+        raise KeyError(f"No implementation found for {cur_mode}.")
+
+    def close(self):
+        for impl in self._singleton.values():
+            impl.close()
+        self._singleton = {}
+
+
+class ProjectKeyService(InterfaceWithLifecycle):
+    @abstractmethod
+    def get_project_key(self, project_id: str, role: ProjectKeyRole) -> Optional[ApiProjectKey]:
+        pass
+
+
+class OrganizationService(InterfaceWithLifecycle):
+    @abstractmethod
+    def get_organizations(
+        self, user_id: Optional[id], scope: Optional[str], only_visible: bool
+    ) -> List[ApiOrganization]:
+        """
+        This method is expected to follow the optionally given user_id, scope, and only_visible options to filter
+        an appropriate set.
+        :param user_id:
+        When null, this should imply the entire set of organizations, not bound by user.  Be certain to authenticate
+        users before returning this.
+        :param scope:
+        :param only_visible:
+        :return:
+        """
+        pass
+
+    @abstractmethod
+    def get_organization_by_slug(
+        self, slug: str, only_visible: bool, allow_stale: bool
+    ) -> Optional[ApiOrganization]:
+        pass
+
+    def _serialize_organization(self, org: Organization) -> ApiOrganization:
+        return ApiOrganization(
+            slug=org.slug,
+            id=org.id,
+        )
+
+
+class DatabaseBackedOrganizationService(OrganizationService):
+    def get_organization_by_slug(
+        self, slug: str, only_visible: bool, allow_stale: bool
+    ) -> Optional[ApiOrganization]:
+        try:
+            if allow_stale:
+                org = Organization.objects.get_from_cache(slug=slug)
+            else:
+                org = Organization.objects.get(slug=slug)
+
+            if only_visible and org.status != OrganizationStatus.VISIBLE:
+                raise Organization.DoesNotExist
+            return self._serialize_organization(org)
+        except Organization.DoesNotExist:
+            logger.info("Active organization [%s] not found", slug)
```

Comment:
Should we-raise a service layer exception when an organization cannot be found instead of returning None?

Reply:
I have been going back and forth on this.  The problem is that future service calls are likely going to be doing batch requests, esp wrt to integration lookups that fetch many underlying models into a synethesized view.  In that case, throwing service level exceptions around the ORM resource design is misleading. 

The other problem is that _honestly that vast majority of callsites just catch the does not exist and expects a None_.  There are many many inline catch blocks that are essentially becoming "exceptions as code flow" all over.  Just forcing a Optional here made more sense because in truth *a non existing model was not an exceptional case, but an expected case*.  

It may become the case that we want to refactor around a common error protocol as we see more of these service layer endpoints come into existence, and in that world we may revisit this.  But for now, I wanted to KISS here.